# Week 5 — Iteration

**COMP 1110 — Introduction to Computer Programming**

**Chapter 5 — Iterations**

Tuesday, October 6 & Thursday, October 8, 2026
9:30 – 11:20 AM · Room OM 1241

Instructor: **Quan Nguyen**

## Where we are

- **Last week (Week 4):** functions — `def`, parameters, return values, scope.
  We learned to *package* a computation and call it by name.
- **This week (Week 5):** **iteration** — `while` and `for`. We learn to *repeat*
  a computation, and the four loop patterns that show up in almost every program
  you will ever write.
- **Next week (Week 6):** Git & GitHub and the final project deliverables on
  Tuesday Oct 13, then **Midterm 1 on Thursday Oct 15** covering Chapters 1–5.
  This week's material is on it.

## Today's outcomes

By the end of these two classes you should be able to:

1. Write a `while` loop with a correct initialisation, condition, and update,
   and explain why all three are required. **[CLO 1]**
2. Write a definite loop with `for` and `range()`, including `start`, `stop`,
   and `step`. **[CLO 1]**
3. Use `break` and `continue`, and recognise the loop-and-a-half pattern.
   **[CLO 1]**
4. Apply the four loop patterns — **counting, summing, maximum, minimum** —
   including the `None`-as-sentinel trick. **[CLO 1, CLO 5]**
5. **Hand-trace** a loop in a table of variable values and predict its output
   without running it. **[CLO 4]**

## Reading

**[Python for Everybody — Chapter 5: Iterations](https://www.py4e.com/html3/05-iterations)**

Read the whole chapter. The Trinket edition is runnable in your browser — edit
every example, change a number, and re-run it.

Pay particular attention to:

- §5.1–5.2 updating variables and the `while` statement
- §5.5 the "loop and a half" with `break`
- §5.7–5.9 counting, summing, maximum and minimum
- §5.10 debugging

Companion videos and practice: <https://www.py4e.com/>

## Updating variables

Everything about loops rests on one idea from Week 2:

```python
x = x + 1
```

This is **not** an equation — it is an instruction. Read it right-to-left:

> "Take the current value of `x`, add 1, and store the result back in `x`."

Names for these:

- `x = x + 1` — an **increment**
- `x = x - 1` — a **decrement**
- `total = total + price` — an **accumulator**

A variable you update this way must be **initialised first**.

In [1]:
# An accumulator: adding up a shopping basket, one item at a time.
total = 0.0                 # initialise BEFORE updating

total = total + 4.99        # milk
print("after milk: ", total)

total = total + 12.50       # cheese
print("after cheese:", total)

total = total + 2.25        # bread
print("after bread: ", total)

print("basket total:", round(total, 2))

after milk:  4.99
after cheese: 17.490000000000002
after bread:  19.740000000000002
basket total: 19.74


Notice `total` is *reassigned* three times. There is only ever one `total`;
each line replaces its value with a new one computed from the old one.

A loop does exactly this — but the update line is written **once** and executed
many times.

## The `while` statement

An **indefinite loop**: we do not know in advance how many times it will run.
It runs *while* a condition stays `True`.

```python
while <boolean expression>:
    <indented body>
```

Every correct `while` loop has **three** parts:

1. **Initialise** the loop variable *before* the loop.
2. **Test** it in the condition.
3. **Update** it *inside* the body, so the condition can eventually become
   `False`.

Miss part 3 and the loop never ends.

In [2]:
# A countdown. Initialise, test, update.
seconds = 3                     # 1. initialise

while seconds > 0:              # 2. test
    print("T minus", seconds)
    seconds = seconds - 1       # 3. update

print("Liftoff!")

T minus 3
T minus 2
T minus 1
Liftoff!


### Hand-tracing that loop

The midterm is closed-book with **no running Python**. So we trace by hand:
one row per pass, one column per variable.

| Pass | `seconds` at test | Test `seconds > 0` | Printed | `seconds` after update |
| :--- | :--- | :--- | :--- | :--- |
| 1 | 3 | `True` | `T minus 3` | 2 |
| 2 | 2 | `True` | `T minus 2` | 1 |
| 3 | 1 | `True` | `T minus 1` | 0 |
| 4 | 0 | `False` | — | loop exits |

Then `print("Liftoff!")` runs once. **Draw this table on paper.** It is the
single most reliable way to answer a trace question.

Two habits worth building now:

- The condition is checked **before** every pass, including the first.
- The loop body may run **zero** times. Set `seconds = 0` above and nothing is
  printed except `Liftoff!`.

### Predict the output (1)

Do not run it. Trace it on paper first.

```python
n = 5
while n > 0:
    print(n)
    n = n - 2
print("Blast off!")
```

What exactly is printed, and how many lines?

**Answer — 4 lines:**

```
5
3
1
Blast off!
```

Trace: `n` is 5 → print, 3 → print, 1 → print, then −1. At the test `-1 > 0` is
`False`, so the loop exits. `n` never equals 0 exactly — which is why
`while n != 0:` would have been an **infinite loop** here. Test with `>`/`<`,
not `!=`, when the step might overshoot.

## Infinite loops — and why we will not run one

An **infinite loop** is a loop whose condition never becomes `False`.

```python
# DO NOT RUN THIS.
n = 10
while n > 0:
    print("Still going...")   # n is never updated!
```

`n` stays 10 forever. The program prints until you kill it. In Jupyter the cell
shows `[*]` and never finishes; the output area grows until the browser tab
freezes.

**How to escape:** press the ⏹ *Interrupt* button (or *Kernel → Interrupt*).
In a terminal, press **Ctrl-C**.

The three ways students actually create one:

1. **No update.** The variable in the condition is never changed in the body.
2. **Update in the wrong direction.** `while n > 0:` with `n = n + 1`.
3. **Update outside the loop.** The update line is not indented into the body,
   so it runs only once, after the loop would have finished.

Before you run a `while` loop, ask out loud: *"what makes this stop?"* If you
cannot answer in one sentence, do not run it.

## `break` — leave the loop now

`break` immediately exits the **innermost** loop. Execution continues at the
first statement after the loop body.

This gives us the **loop-and-a-half**: an intentionally infinite `while True:`
whose only exit is a `break` in the middle. It is the standard shape for
"keep asking until the user says stop".

```python
while True:
    ...get something...
    if <it is time to stop>:
        break
    ...use it...
```

A `while True:` with no `break` is an infinite loop. The `break` *is* the
condition.

In [3]:
# Setup: canned answers so input() never blocks.
def use_canned(*values):
    """Replace input() with one that returns each given value in turn."""
    global input
    remaining = iter(values)

    def canned_input(prompt=""):
        answer = next(remaining)
        print(prompt + answer)
        return answer

    input = canned_input


use_canned("hello", "see you tuesday", "done")

In [4]:
# Echo text messages until the user types 'done'.
while True:
    message = input("Message (or 'done'): ")
    if message == "done":
        break
    print("   sent:", message)

print("Chat closed.")

Message (or 'done'): hello
   sent: hello
Message (or 'done'): see you tuesday
   sent: see you tuesday
Message (or 'done'): done
Chat closed.


## `continue` — skip the rest of *this* pass

`continue` abandons the current pass and jumps straight back to the top of the
loop, re-testing the condition.

- `break` — leave the loop entirely.
- `continue` — skip to the next iteration.

It is how you say *"this item is not interesting, move on"* without wrapping
the whole body in an `else`.

> **Warning:** in a `while` loop, `continue` skips the rest of the body —
> including your update line, if it is at the bottom. That is a classic way to
> write an accidental infinite loop.

In [5]:
use_canned("hi there", "", "", "running late", "done")

In [6]:
# Ignore blank messages; stop at 'done'.
while True:
    message = input("Message (or 'done'): ")
    if message == "":
        continue                     # nothing typed - skip this pass
    if message == "done":
        break
    print("   sent:", message)

print("Chat closed.")

Message (or 'done'): hi there
   sent: hi there
Message (or 'done'): 
Message (or 'done'): 
Message (or 'done'): running late
   sent: running late
Message (or 'done'): done
Chat closed.


### Predict the output (2)

```python
total = 0
for number in range(1, 8):
    if number == 3:
        continue
    if number == 6:
        break
    total = total + number
print(total)
```

What is printed?

**Answer: `12`**

`range(1, 8)` produces 1, 2, 3, 4, 5, 6, 7.

| `number` | what happens | `total` after |
| :--- | :--- | :--- |
| 1 | added | 1 |
| 2 | added | 3 |
| 3 | `continue` — skipped | 3 |
| 4 | added | 7 |
| 5 | added | 12 |
| 6 | `break` — loop ends | 12 |

7 is never reached. `1 + 2 + 4 + 5 = 12`.

## Definite loops: `for` and `range()`

A **definite loop** runs a known number of times. No initialise/test/update to
get wrong — Python handles all three for you.

```python
for <variable> in range(<how many>):
    <indented body>
```

`range(5)` produces the five numbers `0, 1, 2, 3, 4` — it **starts at 0** and
**stops before 5**. That is the single most common source of off-by-one bugs in
this course.

Rule of thumb: `while` when you do not know how many passes; `for` when you do.

In [7]:
# Print the 7 times table.
for multiplier in range(1, 11):
    print("7 x", multiplier, "=", 7 * multiplier)

7 x 1 = 7
7 x 2 = 14
7 x 3 = 21
7 x 4 = 28
7 x 5 = 35
7 x 6 = 42
7 x 7 = 49
7 x 8 = 56
7 x 9 = 63
7 x 10 = 70


In [8]:
# range(start, stop, step): a negative step counts DOWN.
for count in range(5, 0, -1):
    print(count)
print("Go!")

# An empty range: start equals stop, so the body never runs at all.
for count in range(3, 3):
    print("you will never see this")
print("finished")

5
4
3
2
1
Go!
finished


`range(start, stop, step)` — `stop` is **always excluded**:

| Written | Produces |
| :--- | :--- |
| `range(4)` | 0, 1, 2, 3 |
| `range(2, 6)` | 2, 3, 4, 5 |
| `range(0, 20, 5)` | 0, 5, 10, 15 |
| `range(5, 0, -1)` | 5, 4, 3, 2, 1 |
| `range(3, 3)` | *nothing* — the body never runs |

Read `range(a, b)` as *"from `a` up to but not including `b`"* — so
`range(1, 11)` gives 1 through 10, ten numbers.

## The four loop patterns

This is the most valuable half-hour of the course so far. Almost every loop you
write for the rest of the term is one of these four, or a combination:

| Pattern | Initialise before the loop | Update inside the loop |
| :--- | :--- | :--- |
| **Counting** | `count = 0` | `count = count + 1` |
| **Summing** | `total = 0` | `total = total + value` |
| **Maximum** | `largest = None` | `if largest is None or value > largest:` |
| **Minimum** | `smallest = None` | `if smallest is None or value < smallest:` |

Every one has the same shape: **initialise an accumulator, update it once per
item, report it after the loop.**

## Pattern 1 — Counting

*"How many things were there?"*

- Before the loop: `count = 0`
- Inside the loop, once per item: `count = count + 1`
- After the loop: report `count`

Counting only ever cares that an item *happened*, never what its value was.
Put the update inside an `if` and you get **conditional counting** —
"how many days were above 20?"

In [9]:
use_canned("18", "23", "9", "31", "done")

In [10]:
# How many temperatures were entered, and how many were warm (> 20)?
count = 0
warm_days = 0

while True:
    line = input("Temperature (or 'done'): ")
    if line == "done":
        break
    temperature = int(line)
    count = count + 1
    if temperature > 20:
        warm_days = warm_days + 1

print("readings:", count, " warm days:", warm_days)

Temperature (or 'done'): 18
Temperature (or 'done'): 23
Temperature (or 'done'): 9
Temperature (or 'done'): 31
Temperature (or 'done'): done
readings: 4  warm days: 2


**Trace it** for the readings 18, 23, 9, 31:

| Pass | `temperature` | `count` after | warm? | `warm_days` after |
| :--- | :--- | :--- | :--- | :--- |
| 1 | 18 | 1 | no | 0 |
| 2 | 23 | 2 | yes | 1 |
| 3 | 9 | 3 | no | 1 |
| 4 | 31 | 4 | yes | 2 |
| 5 | `"done"` | 4 | — | 2 |

The `"done"` pass breaks *before* `count` is updated — which is exactly why the
`break` is the first thing in the body.

## Pattern 2 — Summing

*"What do they add up to?"*

- Before the loop: `total = 0`
- Inside the loop: `total = total + value`
- After the loop: report `total`

Combine **summing** with **counting** and you get an **average**:
`total / count`.

> Guard the division: if `count` is 0 you get `ZeroDivisionError`. Always ask
> "what if there were no items at all?"

In [11]:
use_canned("18", "23", "9", "31", "done")

In [12]:
# Total and average temperature.
total = 0
count = 0

while True:
    line = input("Temperature (or 'done'): ")
    if line == "done":
        break
    total = total + int(line)
    count = count + 1

if count == 0:
    print("No readings entered.")
else:
    print("total:", total, " average:", round(total / count, 1))

Temperature (or 'done'): 18
Temperature (or 'done'): 23
Temperature (or 'done'): 9
Temperature (or 'done'): 31
Temperature (or 'done'): done
total: 81  average: 20.2


| Pass | value | `total` after | `count` after |
| :--- | :--- | :--- | :--- |
| 1 | 18 | 18 | 1 |
| 2 | 23 | 41 | 2 |
| 3 | 9 | 50 | 3 |
| 4 | 31 | 81 | 4 |

`81 / 4 = 20.25`, rounded to `20.2` (Python rounds half to even).

## Pattern 3 — Maximum, and the `None` sentinel

*"What was the biggest?"*

The tempting start is `largest = 0`. **It is wrong.** If every temperature is
below zero, the answer comes back as `0` — a value that was never in the data.

Instead start with `None`, meaning *"I have not seen anything yet."*

```python
largest = None
...
if largest is None or value > largest:
    largest = value
```

`None` is a value of its own type, `NoneType`. Compare it with `is` / `is not`,
never with `==`.

In [13]:
use_canned("18", "23", "9", "31", "done")

In [14]:
# The highest temperature seen so far.
largest = None

while True:
    line = input("Temperature (or 'done'): ")
    if line == "done":
        break
    temperature = int(line)
    if largest is None or temperature > largest:
        largest = temperature

print("largest:", largest)

Temperature (or 'done'): 18
Temperature (or 'done'): 23
Temperature (or 'done'): 9
Temperature (or 'done'): 31
Temperature (or 'done'): done
largest: 31


| Pass | `temperature` | `largest` before | condition | `largest` after |
| :--- | :--- | :--- | :--- | :--- |
| 1 | 18 | `None` | `is None` → `True` | 18 |
| 2 | 23 | 18 | `23 > 18` → `True` | 23 |
| 3 | 9 | 23 | `9 > 23` → `False` | 23 |
| 4 | 31 | 23 | `31 > 23` → `True` | 31 |

**Why the order matters:** Python's `or` is *short-circuit*. Because
`largest is None` is tested first, `temperature > largest` is never evaluated
on pass 1 — which is what stops the `TypeError` from comparing `int` to `None`.
Swap the two halves and pass 1 crashes.

### Predict the output (3)

```python
largest = None
for number in range(4, 0, -1):
    if largest is None or number > largest:
        largest = number
    print(number, largest)
print("final:", largest)
```

What are the five lines of output?

**Answer:**

```
4 4
3 4
2 4
1 4
final: 4
```

`range(4, 0, -1)` counts **down**: 4, 3, 2, 1 (0 is excluded). `largest` is set
on the very first pass and never beaten again — the biggest value arrived
first. The pattern still works: it does not care about order.

## Pattern 4 — Minimum

Identical to maximum with two characters changed: `>` becomes `<`, and the
variable is renamed.

```python
smallest = None
...
if smallest is None or value < smallest:
    smallest = value
```

The `None` start matters even more here: `smallest = 0` would report `0` for
any data set of positive numbers, and `smallest = 999999` quietly breaks the
day someone enters a bigger number.

In [15]:
use_canned("18", "23", "9", "31", "done")

In [16]:
# Coldest and warmest in one pass - the two patterns side by side.
smallest = None
largest = None

while True:
    line = input("Temperature (or 'done'): ")
    if line == "done":
        break
    temperature = int(line)
    if smallest is None or temperature < smallest:
        smallest = temperature
    if largest is None or temperature > largest:
        largest = temperature

print("coldest:", smallest, " warmest:", largest, " range:", largest - smallest)

Temperature (or 'done'): 18
Temperature (or 'done'): 23
Temperature (or 'done'): 9
Temperature (or 'done'): 31
Temperature (or 'done'): done
coldest: 9  warmest: 31  range: 22


## Debugging loops

When a loop misbehaves, there are only four questions:

1. **Does it start right?** Print the accumulator *before* the loop.
2. **Does it run the right number of times?** Add a counter and print it after.
3. **Does the body do the right thing on pass 1?** Print every variable at the
   top of the body — then delete the print when it works.
4. **Does it stop?** Name the line that makes the condition become `False`.

The universal tool: **put a `print()` inside the loop showing every variable
that changes.** Read the trace like the tables we have been drawing.

In [17]:
# A debugging print, temporarily added to a summing loop.
total = 0
for number in range(1, 6):
    total = total + number
    print("  number =", number, " total =", total)   # remove when fixed

print("final total:", total)

  number = 1  total = 1
  number = 2  total = 3
  number = 3  total = 6
  number = 4  total = 10
  number = 5  total = 15
final total: 15


## Live coding 1 — together

We write this from a blank cell, out loud, as a class.

**Problem.** Ask the user for numbers until they type `done`. Ignore any line
that is not a whole number. Then report how many valid numbers were entered,
their total, their average, the largest, and the smallest.

This is all four patterns in one loop. Start by writing the *initialisations*
before you write anything else.

In [18]:
use_canned("7", "banana", "42", "-3", "done")

In [19]:
# Live coding: read numbers until 'done'.
# Report count, total, average, largest, smallest. Skip bad input.

In [20]:
# Completed solution
count = 0
total = 0
largest = None
smallest = None

while True:
    line = input("Number (or 'done'): ")
    if line == "done":
        break
    try:
        number = int(line)
    except ValueError:
        print("   not a number, ignoring:", line)
        continue
    count = count + 1
    total = total + number
    if largest is None or number > largest:
        largest = number
    if smallest is None or number < smallest:
        smallest = number

if count == 0:
    print("No numbers entered.")
else:
    print("count:", count, " total:", total)
    print("average:", round(total / count, 2))
    print("largest:", largest, " smallest:", smallest)

Number (or 'done'): 7
Number (or 'done'): banana
   not a number, ignoring: banana
Number (or 'done'): 42
Number (or 'done'): -3
Number (or 'done'): done
count: 3  total: 46
average: 15.33
largest: 42  smallest: -3


## Live coding 2 — together

**Problem.** Write a function `count_multiples(limit, factor)` that returns how
many whole numbers from 1 up to and including `limit` are multiples of
`factor`. Then use it to answer: how many numbers from 1 to 100 are multiples
of 7?

Two things to get right: the `range()` bounds, and the counting pattern inside
a function (Week 4 + Week 5 together).

In [21]:
# Live coding: count_multiples(limit, factor)
# How many numbers in 1..limit divide evenly by factor?

In [22]:
# Completed solution
def count_multiples(limit, factor):
    """Return how many numbers in 1..limit are multiples of factor."""
    count = 0
    for number in range(1, limit + 1):
        if number % factor == 0:
            count = count + 1
    return count


print("multiples of 7 up to 100:", count_multiples(100, 7))
print("multiples of 25 up to 100:", count_multiples(100, 25))
print("multiples of 3 up to 10: ", count_multiples(10, 3))

multiples of 7 up to 100: 14
multiples of 25 up to 100: 4
multiples of 3 up to 10:  3


## Common errors this week

**1. `for i in 5:` — you cannot loop over a number.**

```
TypeError: 'int' object is not iterable
```
*Meaning:* `for` needs something that produces items. Write `range(5)`.

**2. Forgetting the update line → the loop never ends.**
No error message at all. The cell shows `[*]` forever. *Fix:* interrupt the
kernel, then find the variable in your condition and make sure the body changes
it.

In [23]:
# Error 1, on purpose.
for i in 5:
    print(i)

TypeError: 'int' object is not iterable

**3. Off-by-one from `range()`.**
`range(1, 10)` gives 1–9, not 1–10. To include `n`, write `range(1, n + 1)`.
*Symptom:* your total or count is short by exactly one item. No error message —
this is a **logic** error, and only tracing finds it.

**4. Comparing to `None` with `==`, or putting the test second.**

```python
if value > largest or largest is None:      # WRONG order
```
```
TypeError: '>' not supported between instances of 'int' and 'NoneType'
```
*Meaning:* the first pass compared a number to `None`. The `is None` test must
come **first** so `or` short-circuits.

## Midterm 1 — Thursday, October 15

**9:30 – 10:30, on Moodle, in class, closed-book.** 25% of your grade.
Covers **Chapters 1–5**. You may **not** run Python.

What to review — one evening per chapter:

- **Ch 1** — what a program is; syntax vs. runtime vs. logic errors.
- **Ch 2** — types, `int()` / `float()` / `str()`, operators, `%`, precedence.
- **Ch 3** — booleans, `and` / `or` / `not`, `if` / `elif` / `else`, nesting,
  `try` / `except`.
- **Ch 4** — `def`, parameters vs. arguments, `return`, local vs. global scope.
- **Ch 5** — `while`, `for`, `range()`, `break`, `continue`, and **all four
  loop patterns**.

**How to prepare (this actually works):**

1. Redo Worksheets 1–4 **on paper, closed-book, timed** (W5 is not due until Oct 20, but do it too). Then check.
2. For every code example in these notebooks, cover the output and predict it.
   Write the trace table.
3. Sit with your project team — they are your study group.
4. Take the **practice quiz** on Moodle. It is the identical interface, so
   nobody meets it for the first time under time pressure.

**Tuesday Oct 13 is the Git & GitHub and project-deliverables class**, so there is
no in-class review. Work through the **Midterm 1 review notebook** on your own
instead, and bring anything you could not answer to office hours.

## Summary

- **Updating a variable** (`x = x + 1`) is the engine of every loop; initialise
  before you update.
- **`while`** is the indefinite loop: initialise, test, **update**. If nothing
  changes the condition, it never stops.
- **`for` + `range()`** is the definite loop. `range(a, b)` stops *before* `b`.
- **`break`** leaves the loop; **`continue`** skips to the next pass.
  `while True:` + `break` is the loop-and-a-half.
- **Four patterns:** counting (`count = 0`), summing (`total = 0`), maximum and
  minimum (start at `None`, test `is None` **first**).
- **Trace before you run.** The table of variable values is the midterm skill.

## Worksheet 5

**Released:** Tuesday, October 13 at 9:00 AM (on the **Practice Platform**).
**Due:** Monday, **October 19** at 11:59 PM — Midterm 1 falls in between, so
finish it before Thursday and use it to revise.

What it covers:

- Writing `while` loops with a correct stopping condition.
- `for` loops over `range()` with `start`, `stop`, and `step`.
- All four loop patterns, plus one that combines counting and summing.
- One **trace-and-explain** question: a hand-drawn table of variable values.
- One **find-the-bug** question: an accidental infinite loop.

Late work: up to 48 hours at −50%. Your lowest worksheet score is dropped.

## Next class

**Tuesday, October 13 — Git & GitHub, and the final project deliverables.**
No prior Git experience is assumed. The Midterm 1 review notebook is yours to
work through on your own.

**Thursday, October 15 — Midterm 1**, 9:30–10:30 on Moodle. TRU login
working, arrive on time. If something fails mid-exam, raise your hand
immediately.

*Before Tuesday:* finish Chapter 5, and do the Moodle practice quiz.